scipy/signal/tests/test_signaltools.py

```diff
@@ -2800,6 +2800,188 @@ def test_hilbert2_types(self, dtype):
+        [([1, 0, 2, 2, 0],    (1, None), [4, 2, 0, 0, 0], [1, 0, 0, 0, 0]),
+         ([4, 0, 2, 0, 0],    (0, None), [4, 0, 2, 0, 0], [0, 0, 0, 0, 0]),
+         ([4, 0, 0, 2, 0], (None, None), [4, 0, 0, 2, 0], [0, 0, 0, 0, 0]),
+         ([0, 0, 2, 2, 0],       (1, 3), [2, 0, 0, 0, 0], [0, 0, 0, 2, 0]),
+         ([4, 0, 2, 2, 0],      (-3, 3), [4, 0, 2, 0, 0], [0, 0, 0, 2, 0]),
+         ([4, 0, 3, 4, 0],    (None, 1), [2, 0, 0, 0, 0], [0, 0, 3, 4, 0]),
+         ([4, 0, 3, 4, 0],    (None, 0), [0, 0, 0, 0, 0], [4, 0, 3, 4, 0])])
+    def test_envelope_real_signals(self, Z, bp_in, Ze2_desired, Zr_desired):
+        """Test envelope calculation with real-valued test signals.
+
+        The comparisons are performed in the Fourier space, since it makes evaluating
+        the bandpass filter behavior straightforward. Note that also the squared
+        envelope can be easily calculated by hand, if one recalls that coefficients of
+        a complex-valued Fourier series representing the signal can be directly
+        determined by an FFT and that the absolute square of a Fourier series is again
+        a Fourier series.
+        """
+        z = sp_fft.irfft(Z)
+        ze2, zr = envelope(z, bp_in, residual='all', squared=True)
+        ze2_lp, zr_lp = envelope(z, bp_in, residual='lowpass', squared=True)
+        Ze2, Zr, Ze2_lp, Zr_lp = (sp_fft.rfft(z_) for z_ in (ze2, zr, ze2_lp, zr_lp))
+
+        Ze2_desired = np.array(Ze2_desired).astype(complex)
+        Zr_desired = np.array(Zr_desired).astype(complex)
+        self.assert_close(Ze2, Ze2_desired,
+                          msg="Envelope calculation error (residual='all')")
+        self.assert_close(Zr, Zr_desired,
+                          msg="Residual calculation error (residual='all')")
+
+        if bp_in[1] is not None:
+            Zr_desired[bp_in[1]:] = 0
+        self.assert_close(Ze2_lp, Ze2_desired,
+                          msg="Envelope calculation error (residual='lowpass')")
+        self.assert_close(Zr_lp, Zr_desired,
+                          msg="Residual calculation error (residual='lowpass')")
+
+    @pytest.mark.parametrize(
+        "               Z,        bp_in,         Ze2_desired,         Zr_desired",
+        [([0, 5, 0, 5, 0], (None, None),    [5, 0, 10, 0, 5],    [0, 0, 0, 0, 0]),
+         ([1, 5, 0, 5, 2],      (-1, 2),    [5, 0, 10, 0, 5],    [1, 0, 0, 0, 2]),
+         ([1, 2, 6, 0, 6, 3],   (-1, 2), [0, 6, 0, 12, 0, 6], [1, 2, 0, 0, 0, 3])
+         ])
+    def test_envelope_complex_signals(self, Z, bp_in, Ze2_desired, Zr_desired):
+        """Test envelope calculation with complex-valued test signals.
+
+        We only need to test for the complex envelope here, since the ``Nones``s in the
+        bandpass filter were already tested in the previous test.
+        """
+        z = sp_fft.ifft(sp_fft.ifftshift(Z))
+        ze2, zr = envelope(z, bp_in, residual='all', squared=True)
```

Comment:
Can you also test for squared=False?

Reply:
Consult [comment above](https://github.com/scipy/scipy/pull/20097/files#r1694233553).